Task 5 1 feature 1

Pl-Alex opened the pull request.

Commit:
task-5 Codebase preparation for the web interactions implementation

Commit:
task-5 User singUp implementation

Commit:
task-5 User singUp extended implementation

1937Carrie commented on gradle/libs.versions.toml:
```diff
@@ -47,11 +51,14 @@ lifecycleViewmodel = { module = "androidx.lifecycle:lifecycle-viewmodel-ktx", ve
 
 # UI Tools
 glide = { module = "com.github.bumptech.glide:glide", version.ref = "glide" }
+logging-interceptor = { module = "com.squareup.okhttp3:logging-interceptor", version.ref = "loggingInterceptor" }
```
com.squareup.okhttp3 library has common dependency version so logging-interceptor and okhttp (plain) have common version. In this case it is "okhttp" version reference.

1937Carrie commented on gradle/libs.versions.toml:
A plugin "jetbrains-kotlin-jvm" has to have dependency version the same as all "org.jetbrains.kotlin.*"-libraries. So there (for org.jetbrains.kotlin.jvm) has to be "kotlin" version reference.

1937Carrie commented on app/src/main/AndroidManifest.xml:
```diff
@@ -14,19 +14,20 @@
         android:label="@string/my_profile_activity_app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
+        android:usesCleartextTraffic="true"
```
It need to replace with networkSecurityConfig attribute. More: [link](https://developer.android.com/privacy-and-security/security-config#CleartextTrafficPermitted).

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/data/ApiService.kt:
```diff
@@ -0,0 +1,45 @@
+    @POST("users")
+    suspend fun createUser(
+        @PartMap map: HashMap<String?, RequestBody?>
+    ): Response<CreateUserResponse>
```
Suggest replace your declaration with
```kotlin
@FormUrlEncoded
@POST("users")
suspend fun createUser(
    @Field("email") email:String, @Field("password") password: String
): Response<CreateUserResponse>
```

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/data/ApiService.kt:
```diff
@@ -0,0 +1,45 @@
+    suspend fun editUser(
+        @Path("userId") userId: String,
+        @Header("Authorization") accessToken: String,
+        @Body requestBody: RequestBody,
```
There has to be model of user as data class object.

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/data/models/Model.kt:
Models should be in separated classes.

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/ui/auth/singUp/SingUpFragment.kt:
```diff
@@ -0,0 +1,109 @@
+            DataStoreProvider(requireContext()),
+            MainRepository(ApiService.getInstance())
+        )
+    }
```
This needs to be reworked using DI.

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/data/repository/MainRepository.kt:
MainRepository should give outside android-compatible data type, retrofit-specific stuff should remain in the webapi module.

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/ui/auth/singUp/SingUpFragment.kt:
```diff
@@ -0,0 +1,109 @@
+import com.alexp.datastore.DataStoreProvider
+import com.alexp.textvalidation.validator.base.ValidationResult
+
+class SingUpFragment : BaseFragment<FragmentSingUpBinding>() {
```
Not "Sing" but "Sign".

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/ui/auth/singUp/SingUpViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+    private var job: Job? = null
+
+    fun createUser(email: String, password: String) {
+        job = CoroutineScope(Dispatchers.IO).launch {
```
Can be replaced with using:
```kotlin
viewModelScope.launch(Dispatchers.IO) {}
```

Commit:
task-5 Changes after review (di not implemented yet)

Commit:
task-5 Changes after review (di implemented)

Commit:
task-5 Changes after review (validation changes)

Commit:
task-5 Changes after review (validation changes)

Commit:
task-5 Changes after review
- Implemented okhttp interceptor to add "Bearer" prefix to Authorization token and fix malformed json responses
- Fixed a state asynchronization error in LogInViewModel
- Simplified password requirements

1937Carrie commented on app/src/main/res/layout/fragment_log_in.xml:
Id of view on 151 line is wrong named.

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/presentation/auth/logIn/LogInViewModel.kt:
```diff
@@ -0,0 +1,81 @@
+            updateEmailFormField(_logInFormState.value?.email ?: "")
+            updatePasswordFormField(_logInFormState.value?.password ?: "")
+            return
+        }
```
This function is called "authorize" but this selected action about check credentials data but not about authorization. You need to rework function by extracting the checking of the credentials into separate function and make one more function for complex sequence. Expected logic sequence:
```kotlin
fun checkDataAndAuthorize() {
   checkData()
   authorize()
}

fun checkData()

fun authorize()
```

1937Carrie commented on app/src/main/java/com/alexP/socialnetwork/presentation/auth/logIn/LogInViewModel.kt:
```diff
@@ -0,0 +1,81 @@
+            mainRepository.authorize(email, password)
+                .onSuccess { response ->
+                    _requestState.postValue(RequestState.Success)
+                    dataStore.setAccessToken(response.data.accessToken)
+                    dataStore.setRefreshToken(response.data.refreshToken)
+                    dataStore.setUserId(response.data.user.id)
```
Actions with datastore does not affect for UI appearance so it can be left somewhere in repository class of business logic.

1937Carrie reviewed (approved):
I left comments that I am sure of, but it also seems to me that it is possible to somehow better process the response from the api-repository. It seems to me that it is better to pass RequestState somewhere in the repository implementation, and not in the viewmodel, that is, for example, if the result of the network request is successful, then we return RequestState.Success(apiService.authorize().body). At the moment, the repository returns Kotlin.Result instead of RequestState.
Well, it is logical to assume that if you have already received the result for the network request, then you are processing the state of the network response, and not the state of the request - a mistake in the naming of the state.

Commit:
task-5 Changes after review
- Moved business logic from the view models to the new AuthRepository.kt
- Moved validation logic from SignUpExtendedFragment.kt and SignUpFragment.kt as it was previously done in LogInFragment.kt
